### PR TITLE
planner: preference range scan for pseudo | tidb-test=pr/2382

### LINKED ISSUE #55632: Planner avoid table scans for pseudo stats if index scan exists
## Enhancement
Pseudo stats are unreliable. Preference a matching index scan over table scan for pseudo stats.

### PR DESCRIPTION
**Pull request by terry1purcell**
### What problem does this PR solve?

Issue Number: close #55632

Problem Summary:

### What changed and how does it work?

Improve index usage with pseudo stats.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/sessionctx/variable/tidb_vars.go [hunk 1]
```diff
@@ -1250,7 +1250,8 @@ const (
 	DefOptConcurrencyFactor                 = 3.0
 	DefOptForceInlineCTE                    = false
 	DefOptInSubqToJoinAndAgg                = true
-	DefOptPreferRangeScan                   = false
+	DefOptPreferRangeScan                   = true
+	DefOptPreferRangeThreshold              = 1
```
Comment (qw4990) [inline pkg/sessionctx/variable/tidb_vars.go:1254]: This variable is not used, can we remove it?

**Review by qw4990 — APPROVED** [review]

File: pkg/sessionctx/variable/tidb_vars.go (hunk 1 above)
Comment (terry1purcell) [reply pkg/sessionctx/variable/tidb_vars.go:1254]: done

**Review by time-and-fate — APPROVED** [review]

File: pkg/planner/core/find_best_task.go [hunk 2]
```diff
@@ -1154,8 +1161,13 @@ func skylinePruning(ds *DataSource, prop *property.PhysicalProperty) []*candidat
 				}
 			}
 			if !ranger.HasFullRange(c.path.Ranges, unsignedIntHandle) {
-				preferredPaths = append(preferredPaths, c)
-				hasRangeScanPath = true
+				// Preference plans with equals/IN predicates or where there is more filtering in the index than against the table
+				equalPlan := c.path.EqCondCount > 0 || c.path.EqOrInCondCount > 0
+				indexFilters := len(c.path.TableFilters) < len(c.path.IndexFilters)
+				if preferMerge || ((equalPlan || indexFilters) && prop.IsSortItemEmpty() || c.isMatchProp) {
```
Comment (winoros) [inline pkg/planner/core/find_best_task.go:1167]:
```suggestion
				if preferMerge || (((equalPlan || indexFilters) && prop.IsSortItemEmpty()) || c.isMatchProp) {
```

**Review by winoros — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Review by wjhuang2016 — APPROVED** [review]

File: pkg/sessionctx/variable/tidb_vars.go [hunk 3]
```diff
@@ -1250,7 +1250,7 @@ const (
 	DefOptConcurrencyFactor                 = 3.0
 	DefOptForceInlineCTE                    = false
 	DefOptInSubqToJoinAndAgg                = true
-	DefOptPreferRangeScan                   = false
+	DefOptPreferRangeScan                   = true
```
Comment (yudongusa) [inline pkg/sessionctx/variable/tidb_vars.go:1253]: This needs document change as well.

**Review by yudongusa — APPROVED** [review]
LGTM except that the document changes is needed as well.

File: pkg/sessionctx/variable/tidb_vars.go (hunk 3 above)
Comment (terry1purcell) [reply pkg/sessionctx/variable/tidb_vars.go:1253]: I will open a doc PR

**Pull request merged**